src/main/java/org/apache/commons/lang3/concurrent/LazyInitializerWithDisposer.java

```diff
@@ -0,0 +1,294 @@
+ * ...
+ * // When the object is actually needed:
+ * ComplexObject cobj = initializer.get();
+ * 
+ * // When it is time to dispose of the object
+ * initializer.dispose();
+ * </pre>
+ *
+ * <p>
+ * If multiple threads call the {@code get()} method when the object has not yet
+ * been created, they are blocked until initialization completes. The algorithm
+ * guarantees that only a single instance of the wrapped object class is
+ * created, which is passed to all callers. Once initialized, calls to the
+ * {@code get()} method are pretty fast because no synchronization is needed
+ * (only an access to a <b>volatile</b> member field).
+ * </p>
+ *
+ * @since 3.14.0
+ * @param <T> the type of the object managed by this initializer class
+ */
+public class LazyInitializerWithDisposer<T> implements ConcurrentInitializer<T> {
+
+    //NO_INIT serves double duty as the lock object to prevent any other class acquiring the monitor
+    private final Object NO_INIT = new Object(){};
+    private final Object DISPOSED = new Object(){};
+
+    private FailableConsumer<? super T, ? extends Exception> disposer;
+    private FailableSupplier<? extends T, ? extends Exception> initializer;
+
+    // Stores the managed object.
+    private volatile T object = (T) NO_INIT;
+
+    private int allowedTries;
+    Exception firstFailure = null;
+
+    /** 
+     * Constructs a LazyInitializerWithDisposer with a given initializer and disposer
+     *
+     * @param initializer an implimentation of the FailableSupplier functional interface which will create the wrapped object
+     * @param disposer an implimentation of the FailableConsumer functional interface which will dispose of the wrapped object
+     * @param allowedTries how many calls to get() will be allowed to attempt to initialize in total, before a failure is cached and becomes persistent. Set to a negative number for infinite retries.
+     */
+    public LazyInitializerWithDisposer(FailableSupplier<? extends T, ? extends Exception> initializer, FailableConsumer<? super T, ? extends Exception> disposer, int allowedTries) {
+        if (allowedTries == 0) {
+           throw new IllegalArgumentException("allowedTries must be a positive or negative number");
+        }
+        this.allowedTries = allowedTries;
+
+        this.initializer = initializer;
+        this.disposer = disposer;
+    }
+
+    /** 
+     * Constructs a LazyInitializerWithDisposer wtih a given initializer and disposer
+     *
+     * @param initializer an implimentation of the FailableSupplier functional interface which will create the wrapped object
+     * @param disposer an implimentation of the FailableConsumer functional interface which will dispose of the wrapped object
+     */
+    public LazyInitializerWithDisposer(FailableSupplier<? extends T, ? extends Exception> initializer, FailableConsumer<? super T, ? extends Exception> disposer) {
+        this(initializer, disposer, 4);
```

Comment:
4 seems random to me. It's not clear to me why this would be needed. Closable and AutoCloseable don't support anything like that for example.

Reply:
There is a logic to that 4. Three retries is a standard default. And I'm using a variable representing tries rather than retries to distinguish between 0 tries left - a fail state - and -1 or below representing infinite tries.

As for why have retries at all. LazyInitializers are most useful when it is expensive to initialize something. Needing some form of IO is both a common reason for an operation to be expensive, and a case where trying again after a failure might actually work. So I think it is useful to make retries possible.